lib/loadObj.js

```diff
@@ -195,24 +296,196 @@ function loadObj(objPath, options) {
-        var index3 = addVertex(v3, p3, u3, n3);
+    function get3DPoint(index, result) {
+        var pi = getOffset(index, positions, 3);
+        var px = positions.get(pi + 0);
+        var py = positions.get(pi + 1);
+        var pz = positions.get(pi + 2);
+        return Cartesian3.fromElements(px, py, pz, result);
+    }
+
+    function get3DNormal(index, result) {
+        var ni = getOffset(index, normals, 3);
+        var nx = normals.get(ni + 0);
+        var ny = normals.get(ni + 1);
+        var nz = normals.get(ni + 2);
+        return Cartesian3.fromElements(nx, ny, nz, result);
+    }
 
-        primitive.indices.push(index1);
-        primitive.indices.push(index2);
-        primitive.indices.push(index3);
+    // Given a sequence of three points A B C, determine whether vector BC
+    // "turns" clockwise (positive) or counter-clockwise (negative) from vector AB
+    var scratch1 = new Cartesian3();
+    var scratch2 = new Cartesian3();
+    function getTurnDirection(pointA, pointB, pointC) {
+        var vector1 = Cartesian2.subtract(pointA, pointB, scratch1);
+        var vector2 = Cartesian2.subtract(pointC, pointB, scratch2);
+        return vector1.x * vector2.y - vector1.y * vector2.x;
+    }
 
-        // Triangulate if the face is a quad
-        if (defined(v4)) {
-            var index4 = addVertex(v4, p4, u4, n4);
+    // Given the cartesian 2 vertices of a polygon, determine if convex
+    function isConvex(positions2D) {
+       var i;
+       var turnDirection = getTurnDirection(positions2D[0], positions2D[1], positions2D[2]);
+       for (i=1; i < positions2D.length-2; ++i) {
+           var currentTurnDirection = getTurnDirection(positions2D[i], positions2D[i+1], positions2D[i+2]);
+            if (turnDirection * currentTurnDirection < 0) {
+                return false;
+            }
+       }
+       return true;
+    }
+
+    var scratch3 = new Cartesian3();
+    // Checks if winding order matches the given normal.
+    function checkWindingCorrect(positionIndex1, positionIndex2, positionIndex3, normal) {
+        var A = get3DPoint(positionIndex1, scratch1);
+        var B = get3DPoint(positionIndex2, scratch2);
+        var C = get3DPoint(positionIndex3, scratch3);
+
+        Cartesian3.subtract(B, A, B);
+        Cartesian3.subtract(C, A, C);
+        var cross = Cartesian3.cross(A, C, scratch1);
+
+        return (Cartesian3.dot(normal, cross) >= 0);
+    }
+
+    function addTriangle(index1, index2, index3, correctWinding) {
+        if (correctWinding) {
+            primitive.indices.push(index1);
+            primitive.indices.push(index2);
+            primitive.indices.push(index3);
+        } else {
             primitive.indices.push(index1);
             primitive.indices.push(index3);
-            primitive.indices.push(index4);
+            primitive.indices.push(index2);
         }
     }
 
+    var scratchNormal = new Cartesian3();
+    function addFace(vertices, positions, uvs, normals) {
```

Comment:
Definitely doable, but the file logic will be pretty tangled though, because it will essentially be all the logic from addFace, which uses addVertex, the global normals / position buffers and the face position and normal buffers. Don't know if it's worth it?